src/useLogin.tsx

```diff
@@ -24,13 +33,52 @@ const attemptLogin = async (code: string): Promise<LoggedInUserData> => {
   }
 }
 
-function useLogin() {
+function useLogin(): UseLoginReturn {
   const firstUpdate = useRef(true)
   const sessionData = useUserSessionDataState()
   const sessionUpdateFn = useUserSessionDataDispatch()
   const [redirect, setRedirect] = React.useState<string | undefined>()
   const [token, setToken] = React.useState(sessionData.token)
   const code = getCode()
+
+  const [isLoadingLoginWithUsernameAndPassword, setIsLoadingLoginWithUsernameAndPassword] =
+    React.useState<boolean>(false)
+  const [errorMessageLoginWithUsernameAndPassword, setErrorMessageLoginWithUsernameAndPassword] = React.useState<
+    string | undefined
+  >()
+
+  const finishLogin = useCallback(
+    async (loginResponse: LoggedInUserData, loginMethod: LoginMethod) => {
+      sessionUpdateFn({
+        type: 'LOGIN',
+        payload: {
+          synapseUserId: loginResponse.synapseUserId,
+          token: loginResponse.sessionToken,
+          firstName: loginResponse.firstName,
+          lastName: loginResponse.lastName,
+          username: loginResponse.username,
+          orgMembership: loginResponse.orgMembership,
+          dataGroups: loginResponse.dataGroups,
+          roles: loginResponse.roles,
+          id: loginResponse.id,
+          appId: Utility.getAppId(),
+          demoExternalId: loginResponse.clientData?.demoExternalId,
+          isVerified: loginResponse.isVerified,
+          lastLoginMethod: loginMethod,
+        },
+      })
+      setToken(loginResponse.sessionToken)
+      const savedLocation = sessionStorage.getItem('location')
+      if (savedLocation) {
+        sessionStorage.removeItem('location')
+        setRedirect(savedLocation)
+      } else {
+        setRedirect('/studies')
+      }
+    },
+    [sessionUpdateFn]
+  )
```

Comment:
Abstracted from the `useEffect` that watches for the OAuth code, so it can be reused when logging with username/password.